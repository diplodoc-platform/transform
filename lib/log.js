const {green, yellow, red} = require('chalk');

const LOG_LEVELS = {
    DISABLED: 'disabled',
    INFO: 'info',
    WARN: 'warn',
    ERROR: 'error',
};

const problems = {
    [LOG_LEVELS.INFO]: [],
    [LOG_LEVELS.WARN]: [],
    [LOG_LEVELS.ERROR]: [],
};

function createLogger(type) {
    const formatter = {
        info: (msg) => `${green('INFO')} ${msg}`,
        warn: (msg) => `${yellow('WARN')} ${msg}`,
        error: (msg) => `${red('ERR ')} ${msg}`,
    };

    return function log(msg) {
        const problem = formatter[type](msg);

        if (!problems[type].includes(problem)) {
            problems[type].push(problem);
        }
    };
}

module.exports = {
    [LOG_LEVELS.INFO]: createLogger(LOG_LEVELS.INFO),
    [LOG_LEVELS.WARN]: createLogger(LOG_LEVELS.WARN),
    [LOG_LEVELS.ERROR]: createLogger(LOG_LEVELS.ERROR),
    LOG_LEVELS,
    get: () => problems,
    clear: () => {
        problems[LOG_LEVELS.INFO] = [];
        problems[LOG_LEVELS.WARN] = [];
        problems[LOG_LEVELS.ERROR] = [];
    },
    isEmpty: () => {
        return !(
            problems[LOG_LEVELS.INFO].length ||
            problems[LOG_LEVELS.WARN].length ||
            problems[LOG_LEVELS.ERROR].length
        );
    },
};
