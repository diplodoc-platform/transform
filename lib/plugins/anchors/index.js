const slugify = require('slugify');
const {bold} = require('chalk');

const {headingInfo} = require('../../utils');
const {CUSTOM_ID_REGEXP} = require('./constants');

function createLinkTokens(state, id, setId = false) {
    const open = new state.Token('link_open', 'a', 1);
    const text = new state.Token('text', '', 0);
    const close = new state.Token('link_close', 'a', -1);

    if (setId) {
        open.attrSet('id', id);
    }
    open.attrSet('href', '#' + id);
    open.attrSet('class', 'yfm-anchor');
    open.attrSet('aria-hidden', 'true');
    text.content = '';

    return [open, text, close];
}

const getCustomIds = (content) => {
    const ids = [];

    content.replace(CUSTOM_ID_REGEXP, (match, customId) => {
        ids.push(customId);
    });

    return ids.length ? ids : null;
};
const removeCustomId = (content) => {
    if (CUSTOM_ID_REGEXP.test(content)) {
        return content.replace(CUSTOM_ID_REGEXP, '').trim();
    }
    return content;
};
const removeCustomIds = (token) => {
    token.content = removeCustomId(token.content);
    token.children.forEach((child) => {
        child.content = removeCustomId(child.content);
    });
};

function index(md, {extractTitleOption, path, log}) {
    const plugin = (state) => {
        /* Do not use the plugin if it is included in the file */
        if (state.env.includes && state.env.includes.length) {
            return;
        }

        const ids = {};
        const tokens = state.tokens;
        let i = 0;

        while (i < tokens.length) {
            const token = tokens[i];
            const isHeading = token.type === 'heading_open';

            if (isHeading) {
                const {title, level} = headingInfo(tokens, i);
                const inlineToken = tokens[i + 1];
                let id = token.attrGet('id');
                let customIds;

                if (!title) {
                    log.warn(`Header without title${path ? ` in ${bold(path)}` : ''}`);
                }

                if (level < 2 && extractTitleOption) {
                    i += 3;
                    continue;
                }

                if (!id) {
                    customIds = getCustomIds(inlineToken.content);
                    if (customIds) {
                        id = customIds[0];
                        removeCustomIds(tokens[i + 1]);
                    } else {
                        id = slugify(title || '', {
                            lower: true,
                            remove: /[*+~.()'"!:@]/g,
                        });
                    }

                    token.attrSet('id', id);
                }

                if (ids[id]) {
                    id = id + ids[id]++;
                    token.attrSet('id', id);
                } else {
                    ids[id] = 1;
                }

                const allAnchorIds = customIds ? customIds : [id];

                allAnchorIds.forEach((customId) => {
                    const setId = id !== customId;
                    const linkTokens = createLinkTokens(state, customId, setId);

                    inlineToken.children.unshift(...linkTokens);
                });

                i += 3;
                continue;
            }

            i++;
        }
    };

    try {
        md.core.ruler.after('includes', 'anchors', plugin);
    } catch (e) {
        md.core.ruler.push('anchors', plugin);
    }
}

module.exports = index;
