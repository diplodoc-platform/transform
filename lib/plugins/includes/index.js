const {bold} = require('chalk');

const {getFileTokens, getFullIncludePath} = require('../../utilsFS');
const {findBlockTokens} = require('../../utils');

const INCLUDE_REGEXP = /^{%\s*include\s*(notitle)?\s*\[(.+?)]\((.+?)\)\s*%}$/;

function stripTitleTokens(tokens) {
    if (tokens[0].type === 'heading_open' && tokens[2].type === 'heading_close') {
        tokens.splice(0, 3);
    }
}

function unfoldIncludes(state, path, options) {
    const {root, notFoundCb, log} = options;
    const {tokens} = state;
    let i = 0;

    while (i < tokens.length) {
        const openToken = tokens[i];
        const contentToken = tokens[i + 1];
        const closeToken = tokens[i + 2];
        let match;

        if (
            openToken.type === 'paragraph_open' &&
            contentToken.type === 'inline' &&
            (match = contentToken.content.match(INCLUDE_REGEXP)) &&
            closeToken.type === 'paragraph_close'
        ) {
            try {
                const [, keyword, /* description */, includePath] = match;
                const fullIncludePath = getFullIncludePath(includePath, root, path);
                const [pathname, hash] = fullIncludePath.split('#');

                if (!pathname.startsWith(root)) {
                    i++;

                    continue;
                }

                const fileTokens = getFileTokens(pathname, state, options);

                let includedTokens;
                if (hash) {
                    includedTokens = findBlockTokens(fileTokens, hash);
                } else {
                    includedTokens = fileTokens;
                }

                if (keyword === 'notitle') {
                    stripTitleTokens(includedTokens);
                }

                tokens.splice(i, 3, ...includedTokens);

                i += includedTokens.length;
            } catch (e) {
                if (notFoundCb) {
                    notFoundCb(e.path.replace(root, ''));
                }
                log.error(`Skip error: ${e}`);

                i++;
            }
        } else {
            i++;
        }
    }
}

function index(md, options) {
    const {path: optPath, log} = options;

    const plugin = (state) => {
        const {env} = state;
        const path = env.path || optPath;

        env.includes = env.includes || [];

        if (env.includes.includes(path)) {
            log.error(`Circular includes: ${bold(env.includes.concat(path).join(' â–¶ '))}`);
            process.exit(1);
        }

        env.includes.push(path);
        unfoldIncludes(state, path, options);
        env.includes.pop();
    };

    try {
        md.core.ruler.before('curly_attributes', 'includes', plugin);
    } catch (e) {
        md.core.ruler.push('includes', plugin);
    }
}

module.exports = index;
