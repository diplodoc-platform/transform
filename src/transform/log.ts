import {green, red, yellow} from 'chalk';

export enum LogLevels {
    INFO = 'info',
    WARN = 'warn',
    ERROR = 'error',
    DISABLED = 'disabled',
}

const problems: Record<LogLevels, string[]> = {
    [LogLevels.INFO]: [],
    [LogLevels.WARN]: [],
    [LogLevels.ERROR]: [],
    [LogLevels.DISABLED]: [],
};

function createLogger(type: LogLevels) {
    const formatter: Record<string, (v: string) => string> = {
        [LogLevels.INFO]: (msg) => `${green('INFO')} ${msg}`,
        [LogLevels.WARN]: (msg) => `${yellow('WARN')} ${msg}`,
        [LogLevels.ERROR]: (msg) => `${red('ERR')} ${msg}`,
    };

    return function log(msg: string) {
        const problem = formatter[type](msg);

        if (!problems[type].includes(problem)) {
            problems[type].push(problem);
        }
    };
}

export const log = {
    [LogLevels.INFO]: createLogger(LogLevels.INFO),
    [LogLevels.WARN]: createLogger(LogLevels.WARN),
    [LogLevels.ERROR]: createLogger(LogLevels.ERROR),
    LogLevels,
    get: () => problems,
    add: (logs: Record<LogLevels, string[]>) => {
        problems[LogLevels.INFO] = problems[LogLevels.INFO].concat(logs[LogLevels.INFO]);
        problems[LogLevels.WARN] = problems[LogLevels.WARN].concat(logs[LogLevels.WARN]);
        problems[LogLevels.ERROR] = problems[LogLevels.ERROR].concat(logs[LogLevels.ERROR]);
    },
    clear: () => {
        problems[LogLevels.INFO] = [];
        problems[LogLevels.WARN] = [];
        problems[LogLevels.ERROR] = [];
    },
    isEmpty: () => {
        return !(
            problems[LogLevels.INFO].length ||
            problems[LogLevels.WARN].length ||
            problems[LogLevels.ERROR].length
        );
    },
};

export type Logger = typeof log;

export default log;
