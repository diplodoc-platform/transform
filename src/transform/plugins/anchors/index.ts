import {bold} from 'chalk';
import GithubSlugger from 'github-slugger';

import {headingInfo} from '../../utils';
import {CUSTOM_ID_REGEXP, CUSTOM_ID_EXCEPTION} from './constants';
import StateCore from 'markdown-it/lib/rules_core/state_core';
import Token from 'markdown-it/lib/token';
import {MarkdownItPluginCb} from '../typings';

const slugify: (str: string, opts: {}) => string = require('slugify');

function createLinkTokens(state: StateCore, id: string, title: string, setId = false) {
    const open = new state.Token('link_open', 'a', 1);
    const close = new state.Token('link_close', 'a', -1);

    if (setId) {
        open.attrSet('id', id);
    }
    open.attrSet('href', '#' + id);
    open.attrSet('class', 'yfm-anchor');
    open.attrSet('aria-hidden', 'true');

    // SEO: render invisible heading title because link must have text content.
    const spanOpen = new state.Token('span_open', 'span', 1);
    const spanText = new state.Token('text', '', 0);
    const spanClose = new state.Token('span_close', 'span', -1);
    spanOpen.attrSet('class', 'visually-hidden');
    spanText.content = title;

    return [open, spanOpen, spanText, spanClose, close];
}

const getCustomIds = (content: string) => {
    const ids: string[] = [];

    content.replace(CUSTOM_ID_REGEXP, (match, customId) => {
        if (match !== CUSTOM_ID_EXCEPTION) {
            ids.push(customId);
        }

        return '';
    });

    return ids.length ? ids : null;
};

const removeCustomId = (content: string) => {
    if (CUSTOM_ID_REGEXP.test(content)) {
        return content
            .replace(CUSTOM_ID_REGEXP, (match) => {
                if (match === CUSTOM_ID_EXCEPTION) {
                    return match;
                }

                return '';
            })
            .trim();
    }

    return content;
};
const removeCustomIds = (token: Token) => {
    token.content = removeCustomId(token.content);
    token.children?.forEach((child) => {
        child.content = removeCustomId(child.content);
    });
};

interface Options {
    extractTitle?: boolean;
    supportGithubAnchors?: boolean;
}

const index: MarkdownItPluginCb<Options> = (
    md,
    {extractTitle, path, log, supportGithubAnchors},
) => {
    const plugin = (state: StateCore) => {
        /* Do not use the plugin if it is included in the file */
        if (state.env.includes && state.env.includes.length) {
            return;
        }

        const ids: Record<string, number> = {};
        const tokens = state.tokens;
        let i = 0;
        const slugger = new GithubSlugger();

        while (i < tokens.length) {
            const token = tokens[i];
            const isHeading = token.type === 'heading_open';

            if (isHeading) {
                const {title, level} = headingInfo(tokens, i);
                const inlineToken = tokens[i + 1];
                let id = token.attrGet('id');
                let ghId: string;

                if (!title) {
                    log.warn(`Header without title${path ? ` in ${bold(path)}` : ''}`);
                }

                if (level < 2 && extractTitle) {
                    i += 3;
                    continue;
                }

                const customIds = getCustomIds(inlineToken.content);
                if (customIds) {
                    id = customIds[0];
                    removeCustomIds(tokens[i + 1]);
                } else {
                    id = slugify(title, {
                        lower: true,
                        remove: /[*+~.()'"!:@`ÑŒ]/g,
                    });
                    ghId = slugger.slug(title);
                }

                token.attrSet('id', id);

                if (ids[id]) {
                    id = id + ids[id]++;
                    token.attrSet('id', id);
                } else {
                    ids[id] = 1;
                }

                const allAnchorIds = customIds ? customIds : [id];
                const anchorTitle = removeCustomId(title).replace(/`/g, '');
                allAnchorIds.forEach((customId) => {
                    const setId = id !== customId;
                    const linkTokens = createLinkTokens(state, customId, anchorTitle, setId);

                    inlineToken.children?.unshift(...linkTokens);

                    if (supportGithubAnchors) {
                        const ghLinkTokens = createLinkTokens(state, ghId, anchorTitle, true);
                        inlineToken.children?.unshift(...ghLinkTokens);
                    }
                });

                i += 3;
                continue;
            }

            i++;
        }
    };

    try {
        md.core.ruler.after('includes', 'anchors', plugin);
    } catch {
        try {
            md.core.ruler.after('curly_attributes', 'anchors', plugin);
        } catch {
            md.core.ruler.push('anchors', plugin);
        }
    }
};

export = index;
