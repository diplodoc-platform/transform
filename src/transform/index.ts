import {bold} from 'chalk';
import attrs from 'markdown-it-attrs';
import Token from 'markdown-it/lib/token';

import {log, LogLevels} from './log';
import makeHighlight from './highlight';
import extractTitle from './title';
import getHeadings from './headings';
import liquid from './liquid';
import sanitizeHtml, {SanitizeOptions} from './sanitize';

import notes from './plugins/notes';
import anchors from './plugins/anchors';
import code from './plugins/code';
import cut from './plugins/cut';
import deflist from './plugins/deflist';
import term from './plugins/term';
import file from './plugins/file';
import imsize from './plugins/imsize';
import meta from './plugins/meta';
import sup from './plugins/sup';
import tabs from './plugins/tabs';
import video from './plugins/video';
import monospace from './plugins/monospace';
import yfmTable from './plugins/table';
import {initMd} from './md';
import {MarkdownItPluginCb} from './plugins/typings';
import type {HighlightLangMap, Heading} from './typings';

interface OutputType {
    result: {
        html: string;
        title?: string;
        headings: Heading[];
        assets?: unknown[];
        meta?: object;
    };
    logs: Record<LogLevels, string[]>;
}
interface OptionsType {
    vars?: Record<string, string>;
    path?: string;
    extractTitle?: boolean;
    needTitle?: boolean;
    allowHTML?: boolean;
    linkify?: boolean;
    breaks?: boolean;
    conditionsInCode?: boolean;
    disableLiquid?: boolean;
    leftDelimiter?: string;
    rightDelimiter?: string;
    isLiquided?: boolean;
    needToSanitizeHtml?: boolean;
    sanitizeOptions?: SanitizeOptions;
    needFlatListHeadings?: boolean;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    plugins?: MarkdownItPluginCb<any>[];
    highlightLangs?: HighlightLangMap;
    root?: string;
    [x: string]: unknown;
}

function transform(originInput: string, opts: OptionsType = {}): OutputType {
    const {
        vars = {},
        path,
        extractTitle: extractTitleOption,
        needTitle,
        allowHTML = false,
        linkify = false,
        breaks = true,
        conditionsInCode = false,
        needToSanitizeHtml = false,
        sanitizeOptions,
        needFlatListHeadings = false,
        disableLiquid = false,
        leftDelimiter = '{',
        rightDelimiter = '}',
        isLiquided = false,
        plugins = [
            meta,
            deflist,
            cut,
            notes,
            anchors,
            tabs,
            code,
            sup,
            video,
            monospace,
            yfmTable,
            file,
            imsize,
            term,
        ],
        highlightLangs = {},
        ...customOptions
    } = opts;

    const pluginOptions = {
        ...customOptions,
        vars,
        path,
        extractTitle: extractTitleOption,
        disableLiquid,
        log,
    };

    const input =
        disableLiquid || isLiquided
            ? originInput
            : liquid(originInput, vars, path, {conditionsInCode});

    const highlight = makeHighlight(highlightLangs);
    const md = initMd({html: allowHTML, linkify, highlight, breaks});
    // Need for ids of headers
    md.use(attrs, {leftDelimiter, rightDelimiter});

    plugins.forEach((plugin) => md.use(plugin, pluginOptions));

    try {
        let title;
        let tokens;
        let titleTokens;
        const env = {} as {[key: string]: Token[] | unknown};

        tokens = md.parse(input, env);

        if (extractTitleOption) {
            ({title, tokens, titleTokens} = extractTitle(tokens));

            // title tokens include other tokens that need to be transformed
            if (titleTokens.length > 1) {
                title = md.renderer.render(titleTokens, md.options, env);
            }
        }
        if (needTitle) {
            ({title} = extractTitle(tokens));
        }

        const headings = getHeadings(tokens, needFlatListHeadings);

        // add all term template tokens to the end of the html
        const termTokens = (env.termTokens as Token[]) || [];
        let html = md.renderer.render([...tokens, ...termTokens], md.options, env);
        if (needToSanitizeHtml) {
            html = sanitizeHtml(html, sanitizeOptions);
        }

        const assets = md.assets;
        const meta = md.meta;

        return {
            result: {html, title, headings, assets, meta},
            logs: log.get(),
        };
    } catch (err) {
        log.error(`Error occurred${path ? ` in ${bold(path)}` : ''}`);
        throw err;
    }
}

export = transform;

// eslint-disable-next-line @typescript-eslint/no-namespace -- backward compatibility
namespace transform {
    export type Options = OptionsType;
    export type Output = OutputType;
}
