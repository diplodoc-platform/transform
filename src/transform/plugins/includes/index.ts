import {bold} from 'chalk';

import {GetFileTokensOpts, getFileTokens, getFullIncludePath, isFileExists} from '../../utilsFS';
import {findBlockTokens} from '../../utils';
import Token from 'markdown-it/lib/token';
import {MarkdownItPluginCb, MarkdownItPluginOpts} from '../typings';
import {StateCore} from 'src/transform/typings';

const INCLUDE_REGEXP = /^{%\s*include\s*(notitle)?\s*\[(.+?)]\((.+?)\)\s*%}$/;

function stripTitleTokens(tokens: Token[]) {
    if (tokens[0].type === 'heading_open' && tokens[2].type === 'heading_close') {
        tokens.splice(0, 3);
    }
}

type Options = MarkdownItPluginOpts &
    GetFileTokensOpts & {
        notFoundCb: (v: string) => void;
        noReplaceInclude: boolean;
    };

function unfoldIncludes(state: StateCore, path: string, options: Options) {
    const {root, notFoundCb, log, noReplaceInclude = false} = options;
    const {tokens} = state;
    let i = 0;

    while (i < tokens.length) {
        const openToken = tokens[i];
        const contentToken = tokens[i + 1];
        const closeToken = tokens[i + 2];
        let match;

        if (
            openToken.type === 'paragraph_open' &&
            contentToken.type === 'inline' &&
            (match = contentToken.content.match(INCLUDE_REGEXP)) &&
            closeToken.type === 'paragraph_close'
        ) {
            try {
                const [, keyword /* description */, , includePath] = match;

                const fullIncludePath = getFullIncludePath(includePath, root, path);
                let pathname = fullIncludePath;
                let hash = '';
                const hashIndex = fullIncludePath.lastIndexOf('#');
                if (hashIndex > -1 && !isFileExists(pathname)) {
                    pathname = fullIncludePath.slice(0, hashIndex);
                    hash = fullIncludePath.slice(hashIndex + 1);
                }

                if (!pathname.startsWith(root)) {
                    i++;

                    continue;
                }

                const fileTokens = getFileTokens(pathname, state, options);

                let includedTokens;
                if (hash) {
                    includedTokens = findBlockTokens(fileTokens, hash);
                } else {
                    includedTokens = fileTokens;
                }

                if (keyword === 'notitle') {
                    stripTitleTokens(includedTokens);
                }

                if (noReplaceInclude) {
                    i++;
                } else {
                    tokens.splice(i, 3, ...includedTokens);

                    i += includedTokens.length;
                }
            } catch (e) {
                // @ts-ignore for some reason typescript fails here
                const errPath = e.path?.replace(root, '');

                if (notFoundCb) {
                    notFoundCb(errPath);
                }
                log.error(`Skip error: ${e} in ${errPath}`);

                i++;
            }
        } else {
            i++;
        }
    }
}

const index: MarkdownItPluginCb<Options> = (md, options) => {
    const {path: optPath, log} = options;

    const plugin = (state: StateCore) => {
        const {env} = state;
        const path = env.path || optPath;

        env.includes = env.includes || [];

        const isCircularInclude = env.includes.includes(path);

        if (isCircularInclude && state.env.disableCircularError) {
            return;
        }

        if (isCircularInclude) {
            log.error(`Circular includes: ${bold(env.includes.concat(path).join(' â–¶ '))}`);
            process.exit(1);
        }

        env.includes.push(path);
        unfoldIncludes(state, path, options);
        env.includes.pop();
    };

    try {
        md.core.ruler.before('curly_attributes', 'includes', plugin);
    } catch (e) {
        md.core.ruler.push('includes', plugin);
    }
};

export = index;
