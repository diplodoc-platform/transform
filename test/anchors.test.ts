import {dirname} from 'path';

import includes from '../src/transform/plugins/includes';
import anchors from '../src/transform/plugins/anchors';
import {log} from '../src/transform/log';
import transform from '../src/transform';

const mocksPath = require.resolve('./utils.ts');
const transformYfm = (text: string) => {
    const {
        result: {html},
    } = transform(text, {
        plugins: [includes, anchors],
        path: mocksPath,
        root: dirname(mocksPath),
    });
    return html;
};

describe('Anchors', () => {
    beforeEach(() => {
        log.clear();
    });

    it('should add single anchor with auto naming', () => {
        expect(transformYfm('## Test\n' + '\n' + 'Content\n')).toBe(
            '<h2 id="test"><a href="#test" class="yfm-anchor" aria-hidden="true"><span class="visually-hidden">Test</span></a>Test</h2>\n' +
                '<p>Content</p>\n',
        );
    });

    it('should add single anchor', () => {
        expect(transformYfm('## Test {#test1}\n' + '\n' + 'Content\n')).toBe(
            '<h2 id="test1"><a href="#test1" class="yfm-anchor" aria-hidden="true"><span class="visually-hidden">Test</span></a>Test</h2>\n' +
                '<p>Content</p>\n',
        );
    });

    it('should add multiple anchors', () => {
        expect(transformYfm('## Test {#test1} {#test2} {#test3}\n' + '\n' + 'Content\n')).toBe(
            '<h2 id="test1">' +
                '<a id="test3" href="#test3" class="yfm-anchor" aria-hidden="true"><span class="visually-hidden">Test</span></a>' +
                '<a id="test2" href="#test2" class="yfm-anchor" aria-hidden="true"><span class="visually-hidden">Test</span></a>' +
                '<a href="#test1" class="yfm-anchor" aria-hidden="true"><span class="visually-hidden">Test</span></a>Test</h2>\n' +
                '<p>Content</p>\n',
        );
    });

    it('should add single anchor when included', () => {
        expect(
            transformYfm(
                '## Test {#test0}\n' +
                    '\n' +
                    'Content before include\n' +
                    '\n' +
                    '{% include [test](./mocks/include-anchor.md) %}\n',
            ),
        ).toBe(
            '<h2 id="test0"><a href="#test0" class="yfm-anchor" aria-hidden="true"><span class="visually-hidden">Test</span></a>Test</h2>\n' +
                '<p>Content before include</p>\n' +
                '<h1 id="test1"><a href="#test1" class="yfm-anchor" aria-hidden="true"><span class="visually-hidden">Title</span></a>Title</h1>\n' +
                '<p>Content</p>\n',
        );
    });

    it('should add multiple anchors when included', () => {
        expect(
            transformYfm(
                '## Test {#test0}\n' +
                    '\n' +
                    'Content before include\n' +
                    '\n' +
                    '{% include [test](./mocks/include-multiple-anchors.md) %}\n',
            ),
        ).toBe(
            '<h2 id="test0"><a href="#test0" class="yfm-anchor" aria-hidden="true"><span class="visually-hidden">Test</span></a>Test</h2>\n' +
                '<p>Content before include</p>\n' +
                '<h1 id="test1">' +
                '<a id="test3" href="#test3" class="yfm-anchor" aria-hidden="true"><span class="visually-hidden">Title</span></a>' +
                '<a id="test2" href="#test2" class="yfm-anchor" aria-hidden="true"><span class="visually-hidden">Title</span></a>' +
                '<a href="#test1" class="yfm-anchor" aria-hidden="true"><span class="visually-hidden">Title</span></a>Title</h1>\n' +
                '<p>Content</p>\n',
        );
    });

    it('should be transliterated correctly', () => {
        expect(transformYfm('## Максимальный размер дисков \n' + '\n' + 'Content\n')).toBe(
            '<h2 id="maksimalnyj-razmer-diskov">' +
                '<a href="#maksimalnyj-razmer-diskov" class="yfm-anchor" aria-hidden="true"><span class="visually-hidden">Максимальный размер дисков</span></a>' +
                'Максимальный размер дисков' +
                '</h2>\n' +
                '<p>Content</p>\n',
        );
    });

    it('should be removed fences after transliteration', () => {
        expect(transformYfm('## `Test`\n' + '\n' + 'Content\n')).toBe(
            '<h2 id="test">' +
                '<a href="#test" class="yfm-anchor" aria-hidden="true"><span class="visually-hidden">Test</span></a><code>Test</code>' +
                '</h2>\n' +
                '<p>Content</p>\n',
        );
    });

    it('should include content by anchor in sharped path file', () => {
        expect(
            transformYfm(
                'Content before include\n' +
                    '\n' +
                    '{% include [file](./mocks/folder-with-#-sharp/file-with-#-sharp.md#anchor) %}\n' +
                    '\n' +
                    'After include',
            ),
        ).toBe(
            '<p>Content before include</p>\n' +
                '<h2 id="anchor"><a href="#anchor" class="yfm-anchor" aria-hidden="true"><span class="visually-hidden">Subtitle</span></a>Subtitle</h2>\n' +
                '<p>Subcontent</p>\n' +
                '<p>After include</p>\n',
        );
    });

    it('should add anchor with auto naming, using entire heading text', () => {
        expect(transformYfm('## _Lorem ~~ipsum **dolor** sit~~ amet_\n\nParagraph\n')).toBe(
            '<h2 id="lorem-ipsum-dolor-sit-amet">' +
                '<a href="#lorem-ipsum-dolor-sit-amet" class="yfm-anchor" aria-hidden="true">' +
                '<span class="visually-hidden">Lorem ipsum dolor sit amet</span></a>' +
                '<em>Lorem <s>ipsum <strong>dolor</strong> sit</s> amet</em></h2>\n' +
                '<p>Paragraph</p>\n',
        );
    });
});
