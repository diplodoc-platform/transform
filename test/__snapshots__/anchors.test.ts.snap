// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Anchors should add anchor with auto naming, using entire heading text 1`] = `
<h2 id="lorem-ipsum-dolor-sit-amet">
  <a href="link.html#lorem-ipsum-dolor-sit-amet"
     class="yfm-anchor"
     aria-hidden="true"
  >
    <span class="visually-hidden"
          data-no-index="true"
    >
      Lorem ipsum dolor sit amet
    </span>
  </a>
  <em>
    Lorem
    <s>
      ipsum
      <strong>
        dolor
      </strong>
      sit
    </s>
    amet
  </em>
</h2>
<p>
  Paragraph
</p>
`;

exports[`Anchors should add id for heading with allowed chars in fragment (RFC 3986)  1`] = `
<h2 id="a-c5percentdollarandabc_;xx,=3/bbb">
  <a href="link.html#a-c5percentdollarandabc_;xx,=3/bbb"
     class="yfm-anchor"
     aria-hidden="true"
  >
    <span class="visually-hidden"
          data-no-index="true"
    >
      A-c~5%!$&amp;(abc)*_+;xx,@=3/':b?b.b'
    </span>
  </a>
  A-c~5%!$&amp;(abc)*_+;xx,@=3/':b?b.b'
</h2>
<p>
  Content
</p>
`;

exports[`Anchors should add multiple anchors 1`] = `
<pre>
  <code>
    ## Test {#test1} {#test2} {#test3}

            Content
  </code>
</pre>
`;

exports[`Anchors should add multiple anchors when included 1`] = `
<h2 id="test0">
  <a href="link.html#test0"
     class="yfm-anchor"
     aria-hidden="true"
  >
    <span class="visually-hidden"
          data-no-index="true"
    >
      Test
    </span>
  </a>
  Test
</h2>
<p>
  Content before include
</p>
<h1 id="test1">
  <a id="test3"
     href="link.html#test3"
     class="yfm-anchor"
     aria-hidden="true"
  >
    <span class="visually-hidden"
          data-no-index="true"
    >
      Title
    </span>
  </a>
  <a id="test2"
     href="link.html#test2"
     class="yfm-anchor"
     aria-hidden="true"
  >
    <span class="visually-hidden"
          data-no-index="true"
    >
      Title
    </span>
  </a>
  <a href="link.html#test1"
     class="yfm-anchor"
     aria-hidden="true"
  >
    <span class="visually-hidden"
          data-no-index="true"
    >
      Title
    </span>
  </a>
  Title
</h1>
<p>
  Content
</p>
`;

exports[`Anchors should add single anchor 1`] = `
<h2 id="test1">
  <a href="link.html#test1"
     class="yfm-anchor"
     aria-hidden="true"
  >
    <span class="visually-hidden"
          data-no-index="true"
    >
      Test
    </span>
  </a>
  Test
</h2>
<p>
  Content
</p>
`;

exports[`Anchors should add single anchor when included 1`] = `
<h2 id="test0">
  <a href="link.html#test0"
     class="yfm-anchor"
     aria-hidden="true"
  >
    <span class="visually-hidden"
          data-no-index="true"
    >
      Test
    </span>
  </a>
  Test
</h2>
<p>
  Content before include
</p>
<h1 id="test1">
  <a href="link.html#test1"
     class="yfm-anchor"
     aria-hidden="true"
  >
    <span class="visually-hidden"
          data-no-index="true"
    >
      Title
    </span>
  </a>
  Title
</h1>
<p>
  Content
</p>
`;

exports[`Anchors should add single anchor with auto naming 1`] = `
<h2 id="test">
  <a href="link.html#test"
     class="yfm-anchor"
     aria-hidden="true"
  >
    <span class="visually-hidden"
          data-no-index="true"
    >
      Test
    </span>
  </a>
  Test
</h2>
<p>
  Content
</p>
`;

exports[`Anchors should be removed fences after transliteration 1`] = `
<h2 id="test">
  <a href="link.html#test"
     class="yfm-anchor"
     aria-hidden="true"
  >
    <span class="visually-hidden"
          data-no-index="true"
    >
      Test
    </span>
  </a>
  <code>
    Test
  </code>
</h2>
<p>
  Content
</p>
`;

exports[`Anchors should be transliterated correctly 1`] = `
<h2 id="maksimalnyj-razmer-diskov">
  <a href="link.html#maksimalnyj-razmer-diskov"
     class="yfm-anchor"
     aria-hidden="true"
  >
    <span class="visually-hidden"
          data-no-index="true"
    >
      Максимальный размер дисков
    </span>
  </a>
  Максимальный размер дисков
</h2>
<p>
  Content
</p>
`;

exports[`Anchors should include content by anchor in sharped path file 1`] = `
<p>
  Content before include
</p>
<h2 id="anchor">
  <a href="link.html#anchor"
     class="yfm-anchor"
     aria-hidden="true"
  >
    <span class="visually-hidden"
          data-no-index="true"
    >
      Subtitle
    </span>
  </a>
  Subtitle
</h2>
<p>
  Subcontent
</p>
<p>
  After include
</p>
`;

exports[`Anchors should remove quotes from id 1`] = `
<h2 id="heading-with-quotes">
  <a href="link.html#heading-with-quotes"
     class="yfm-anchor"
     aria-hidden="true"
  >
    <span class="visually-hidden"
          data-no-index="true"
    >
      «Heading» “with” "quotes'
    </span>
  </a>
  «Heading» “with” "quotes'
</h2>
<p>
  Content
</p>
`;

exports[`Anchors with extract title should not add an anchor for level 1 heading 1`] = `
<p>
  Content
</p>
`;
